### PR TITLE
feat: refactoring casl permission checks for recursive nested operations

### LINKED ISSUE #830: Add nested permission layer on Server
Our ability handlers are not capable of making sure nested updates are respecting workspace scoping. This could lead to security issues.

We are exploring ways to ensure that workspace scoping is always respected

### PR DESCRIPTION
**Pull request by magrinj**
This PR introduces a new capability to the project: checking permissions for Prisma operations involving nested relationships. This new feature ensures that the right permissions are in place when performing complex nested operations in our database.

Key highlights include:

1. **Operations Definition**: Operations are categorized into 'create' and 'simple' based on the type of permission checks they require. This classification improves how we process each operation.

2. **Operations Checking**: We've introduced 'OperationAbilityCheckers' that conduct permission checks depending on the operation type. 'createAbilityCheck' ensures a user has the ability to create a new entity, while 'simpleAbilityCheck' checks permissions for operations like 'connect', 'disconnect', 'delete', etc.

3. **Recursive Nested Operations**: We have introduced a new function, `relationAbilityChecker`, that handles recursive nested operations. It splits each operation and checks them independently.

Here is an example of how to use these new permission checks:

```typescript
@Injectable()
export class CreateCompanyAbilityHandler implements IAbilityHandler {
  constructor(private readonly prismaService: PrismaService) {}

  async handle(ability: AppAbility, context: ExecutionContext) {
    const gqlContext = GqlExecutionContext.create(context);
    const args = gqlContext.getArgs();

    const allowed = await relationAbilityChecker(
      'Company',
      ability,
      this.prismaService.client,
      args,
    );

    if (!allowed) {
      return false;
    }

    return ability.can(AbilityAction.Create, 'Company');
  }
}
```

### CODE REVIEW

File: server/src/ability/ability.util.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
+import { Prisma, PrismaClient } from '@prisma/client';
+import { subject } from '@casl/ability';
+
+import { camelCase } from 'src/utils/camel-case';
+
+import { AppAbility } from './ability.factory';
+import { AbilityAction } from './ability.action';
+
+type OperationType =
```
Comment (lucasbordeau) [inline server/src/ability/ability.util.ts:9]: Can we extract this from prisma ?

Comment (lucasbordeau) [inline server/src/ability/ability.util.ts:9]: Can we extract this from prisma ?

File: server/src/ability/ability.util.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+const simpleAbilityCheck: OperationAbilityChecker = async (
+  modelName,
+  ability,
+  prisma,
+  data,
+) => {
+  // Extract entity name from model name
+  const entity = camelCase(modelName);
+  // Handle all operations cases
+  const operations = !Array.isArray(data) ? [data] : data;
```
Comment (lucasbordeau) [inline server/src/ability/ability.util.ts:59]: Is it the same parsing operation as above ? Could we extract it into a parseDataArray or something like that ?

File: server/src/ability/ability.util.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  return true;
+};
+
+const operationAbilityCheckers: Record<OperationType, OperationAbilityChecker> =
+  {
+    create: createAbilityCheck,
+    createMany: createAbilityCheck,
+    upsert: simpleAbilityCheck,
```
Comment (charlesBochet) [inline server/src/ability/ability.util.ts:85]: I think that we should find a better naming than "simple". existing?

File: server/src/ability/ability.util.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+) => Promise<boolean>;
+
+const createAbilityCheck: OperationAbilityChecker = async (
+  modelName,
+  ability,
+  prisma,
+  data,
+) => {
+  // Handle all operations cases
+  const items = !Array.isArray(data?.data ?? data) ? [data] : data;
```
Comment (charlesBochet) [inline server/src/ability/ability.util.ts:38]: const operations

**Review by charlesBochet — APPROVED** [review]
Bravo!

**Pull request merged**